include/crm/common/logging.h

```diff
@@ -259,8 +259,17 @@ unsigned int get_crm_log_level(void);
                                         "xml-blob", (level), __LINE__, 0);  \
                 }                                                           \
                 if (crm_is_callsite_active(xml_cs, (level), 0)) {           \
-                    log_data_element((level), __FILE__, __func__,           \
-                         __LINE__, text, xml, 1, xml_log_option_formatted); \
+                    if (xml == NULL) {                                      \
+                        do_crm_log((level), "%s%sNo data to dump as XML",   \
+                                   pcmk__s(text, ""),                       \
+                                   pcmk__str_empty(text)? "" : " ");        \
+                    } else {                                                \
+                        pcmk__xml_log((level), text, xml, 1,                \
+                                      xml_log_option_formatted              \
+                                      |xml_log_option_open                  \
+                                      |xml_log_option_children              \
+                                      |xml_log_option_close);               \
+                    }                                                       \
```

Comment:
pcmk__xml_log() can't be called from a public header. We could make it public, or create a new public function that's essentially a rename of do_crm_log_xml() and make that a deprecated wrapper, but I'd rather not since I'd like to take all of the XML logging out of the public API eventually.

I think we can make do_crm_log_xml() a function instead of macro and anything compiled with the old header will just continue to use the old macro instead of the new function when relinked against the new libraries, but that's worth confirming.

Reply:
Ouch, yep.

> I think we can make do_crm_log_xml() a function instead of macro and anything compiled with the old header will just continue to use the old macro instead of the new function when relinked against the new libraries, but that's worth confirming.

That's what I would *expect*. If it was compiled against the old header, then I would think the new function hasn't been declared and so it's not aware of it. More importantly, that symbol isn't even present in the compiled code, right? The preprocessor should expand the `do_crm_log_xml()` macro into its definition.

I'm also not sure whether it matters or not... if it does use the new function, and the new function behaves the same as the old macro, would that even cause a problem? (Our previous ABI discussion was helpful but there's still a lot I don't know.)

Reply:
The only question is whether an externally compiled application would run at all when the shared library was upgraded, or crash due to a symbol conflict or something like that. I believe defining a macro with the same name as a library API overrides the API since it's effectively stripped out at the preprocessor stage, but I haven't confirmed that.

Reply:
Some time soon I need to figure out how to build a dummy program that links against the Pacemaker libraries so I can test things like this instead of assuming.

I'm not ruling anything out, but I would be fairly surprised if this caused a symbol conflict since (as we both said) the preprocessor expands and strips out the macro in the first step.

Reply:
devel/scratch.c provides a quick template. Just add a call to do_crm_log_xml(), compile the program with the 2.1.4 libraries installed (i.e. older shared object version), run the program to make sure it works, then install your new packages and try to run the program again.

I expect there wouldn't be any problems, but compilers have gotten so picky these days that I don't know if they'd issue a warning for this.